autonat/README.md

```diff
@@ -0,0 +1,127 @@
+# NAT Discovery
+> How we detect if we're behind a NAT.
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Recommendation | Active | r0, 2021-08-26  |
+
+
+Authors: [@marten-seemann]
+
+Interest Group: [@mxinden], [@vyzo], [@raulk], [@stebalien], [@willscott]
+
+[@marten-seemann]: https://github.com/marten-seemann
+[@mxinden]: https://github.com/mxinden/
+[@vyzo]: https://github.com/vyzo
+[@raulk]: https://github.com/raulk
+[@stebalien]: https://github.com/stebalien
+[@willscott]: https://github.com/willscott
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+## Table of Contents
+
+TODO: fill in
+
+## Overview
+
+A priori, a node cannot know if it is behind a NAT / firewall or if it is publicly reachable.
+Knowing its NAT status is essential for the node to be well-behaved in the
+network: A node that's behind a NAT doesn't need to advertise its (undiable)
+addresses to the rest of the network, preventing superfluous dials from other
+peers. Furthermore, it might actively seek to improve its connectivity by
+finding a relay server, which would allow other peers to establish a relayed
+connection.
+
+To determine if it is located behind a NAT, nodes use the `autonat` protocol.
+Using this protocol, the node requests other peers to dial its presumed public
+addresses. If a couple of these dial attempts succeed, the node can be reasonably
+certain that it is not located behind a NAT. Likewise, if a couple of these dial
+attempts fail, this is a strong indicator that a NAT is blocking incoming
+connections.
+
+## AutoNAT Protocol
+
+The AutoNAT Protocol uses the Protocol ID `/libp2p/autonat/1.0.0`. The node
+wishing to determine its NAT status opens a stream using this protocol ID, and
+then sends a `Dial` message. The `Dial` message contains a list of multiaddresses.
```

Comment:
The `PeerInfo` that is send in the `Dial` message also (may?) contain the target's peer ID. Should it be enforced that this Peer ID matches the sender's Peer ID?

Reply:
I would expect addresses in `PeerInfo::addrs` not to contain peer IDs (`/p2p/QmXXX`), given that the peer ID is already carried in `PeerInfo::id`.

That said, I would suggest to take the liberal approach of allowing addresses in `PeerInfo::addrs` to contain peer IDs anyways. Though in that case, I do agree that implementations should enforce the peer IDs to match.

Reply:
> given that the peer ID is already carried in PeerInfo::id

Sorry I probably placed my comment wrong.
I was referring to `PeerInfo::id`, not `PeerInfo::addrs`.

Reply:
Ah, sorry, my bad.

The Golang implementation does the validation you suggest, namely validate that the peer ID in `PeerInfo` matches the remote's peer ID.

https://github.com/libp2p/go-libp2p-autonat/blob/005802887af9a40607fad1283f0f16dbf31e93a9/svc.go#L90-L100

To be honest, I don't know why we send the peer ID in the first place. Maybe @marten-seemann knows more.